File: src/main/java/at/yawk/math/data/Matrix.java
package at.yawk.math.data;

import java.util.List;
import java.util.stream.Collectors;
import lombok.Builder;
import lombok.Singular;
import lombok.Value;

/**
 * @author yawkat
 */
@Value
public class Matrix implements Expression {
    private final List<Vector> columns;
    private final int height;

    @Builder
    private Matrix(@Singular List<Vector> vertices) {
        this.height = vertices.isEmpty() ? 0 : vertices.get(0).getDimension();
        for (Vector vector : vertices) {
            if (vector.getDimension() != height) {
                throw new IllegalArgumentException("Illegal dimension for vector " + vector + ", expected " + height);
            }
        }
        this.columns = vertices;
    }

    public int getWidth() {
        return columns.size();
    }

    @Override
    public String toString(int radix) {
        // ((1, 2, 3), (4, 5, 6))^T
        return columns.stream().map(v -> v.toString(radix))
                .collect(Collectors.joining(", ", "(", ")"));
    }

    @Override
    public String toString() {
        return toString(10);
    }
}

File: src/main/kotlin/at/yawk/math/data/Matrix.kt
package at.yawk.math.data

import at.yawk.math.EqualsHelper

/**
 * @author yawkat
 */
class Matrix(val columns: List<Vector>) : Expression {
    val height: Int
    val width: Int
        get() = columns.size

    init {
        height = columns.firstOrNull()?.dimension ?: 0
        for (col in columns) {
            if (col.dimension != height) {
                throw IllegalArgumentException("Illegal dimension for column $col, expected $height")
            }
        }
    }

    override fun toString(radix: Int): String {
        return columns.map { it.toString(radix) }.joinToString(", ", "(", ")")
    }

    override fun equals(other: Any?): Boolean {
        return EqualsHelper.equals<Matrix>(other, { it.columns == columns })
    }

    override fun hashCode(): Int {
        return EqualsHelper.hashCode(columns)
    }
}